DashWallet/Sources/Models/Taxes/Taxes.swift

```diff
@@ -61,8 +69,16 @@ class Taxes: NSObject {
     func taxCategory(for tx: DSTransaction) -> TxUserInfoTaxCategory {
         var taxCategory: TxUserInfoTaxCategory = tx.defaultTaxCategory()
 
-        if let inputAddress = tx.inputAddresses.first as? String, let txCategory = self.taxCategory(for: inputAddress) {
-            taxCategory = txCategory
+        for outputAddress in tx.outputAddresses {
+            if let address = outputAddress as? String, let txCategory = self.taxCategory(for: address) {
+                // Some transactions might have output that returns change
+                // to the input same address, so need to check that directions match.
+                if (tx.direction() == .sent && txCategory.isOutgoing) ||
+                    (tx.direction() == .received && txCategory.isIncoming) {
+                    taxCategory = txCategory
+                    break
+                }
+            }
```

Comment:
We need to check all addresses because there's no guarantee that the first one will be the one we marked.

Additionally, some transfers (eg CrowdNode) return changes to the same address that can be marked with another category, so we need to make a sanity check to compare the direction of the tax category and transaction direction.

Reply:
@Syn-McJ Sounds good. The only thing here, that I would add is calling `tx.direction()` once and save the result into some variable and then compare the variable to enum cases. I believe this part will be optimazed during the compilation, but it's still better to calculate the direction just once.

Reply:
fixed